presto-main/src/main/java/com/facebook/presto/server/remotetask/TaskInfoFetcher.java

```diff
@@ -96,6 +100,7 @@ public TaskInfoFetcher(
         this.taskId = initialTask.getTaskStatus().getTaskId();
         this.onFail = requireNonNull(onFail, "onFail is null");
         this.taskInfo = new StateMachine<>("task " + taskId, executor, initialTask);
+        this.finalTaskInfo = new StateMachine<>("task-" + taskId, executor, Optional.empty());
```

Comment:
We already track `taskInfo`, why can't/don't we track whether it's final or not instead of having a separate `TaskInfo` instance around (the `finalTaskInfo` instance)?

Reply:
I believe it is possible to combine them into one state machine, but I think this is a simpler design and it matches the design at stage and query level... so it should be easier for people digging into the final info codebase when they all look the same.  Anyway, that is what I was thinking.